source/common/network/apple_dns_impl.cc

```diff
@@ -113,22 +113,55 @@ ActiveDnsQuery* AppleDnsResolverImpl::resolve(const std::string& dns_name,
   ENVOY_LOG(debug, "DNS resolver resolve={}", dns_name);
-  std::unique_ptr<PendingResolution> pending_resolution(
-      new PendingResolution(*this, callback, dispatcher_, main_sd_ref_, dns_name));
 
-  DNSServiceErrorType error = pending_resolution->dnsServiceGetAddrInfo(dns_lookup_family);
-  if (error != kDNSServiceErr_NoError) {
-    ENVOY_LOG(warn, "DNS resolver error ({}) in dnsServiceGetAddrInfo for {}", error, dns_name);
-    return nullptr;
+  Address::InstanceConstSharedPtr address{};
+  try {
+    // After production testing with this resolver implementation it became empirically demonstrable
+    // that Apple's API issues queries to the DNS server for dns_name that might already be an IP
+    // address. In contrast, c-ares synchronously resolves such cases. Moreover, some DNS servers
+    // might not resolve IP addresses and thus result in callback targets that never get a
+    // resolution. Therefore, we short circuit here by trying to parse the dns_name into an internet
+    // address and synchronously issue the ResolveCb; only if the parsing throws an exception the
+    // resolver issues a call to Apple's API.
+    address = Utility::parseInternetAddress(dns_name);
+    ENVOY_LOG(debug, "DNS resolver resolved ({}) to ({}) without issuing call to Apple API",
+              dns_name, address->asString());
+  } catch (const EnvoyException& e) {
+    // Resolution via Apple APIs
+    ENVOY_LOG(debug, "DNS resolver local resolution failed with: {}", e.what());
+    std::unique_ptr<PendingResolution> pending_resolution(
+        new PendingResolution(*this, callback, dispatcher_, main_sd_ref_, dns_name));
+
+    DNSServiceErrorType error = pending_resolution->dnsServiceGetAddrInfo(dns_lookup_family);
+    if (error != kDNSServiceErr_NoError) {
+      ENVOY_LOG(warn, "DNS resolver error ({}) in dnsServiceGetAddrInfo for {}", error, dns_name);
+      return nullptr;
+    }
+
+    // If the query was synchronously resolved in the Apple API call, there is no need to return the
+    // query.
+    if (pending_resolution->synchronously_completed_) {
+      return nullptr;
+    }
+
+    pending_resolution->owned_ = true;
+    return pending_resolution.release();
   }
 
-  // If the query was synchronously resolved, there is no need to return the query.
-  if (pending_resolution->synchronously_completed_) {
+  ASSERT(address != nullptr);
+  // Finish local, synchronous resolution. This needs to happen outside of the exception block above
+  // as the callback itself can throw.
+  try {
+    callback(DnsResolver::ResolutionStatus::Success,
+             {DnsResponse(address, std::chrono::seconds(60))});
     return nullptr;
+  } catch (const std::exception& e) {
+    ENVOY_LOG(warn, "std::exception in callback with local resolution: {}", e.what());
+    throw EnvoyException(e.what());
+  } catch (...) {
+    ENVOY_LOG(warn, "Unknown exception in callback with local resolution");
+    throw EnvoyException("unknown");
```

Comment:
Why is this try catch needed vs. just letting the callback through if that is what it does?

Reply:
I was imitating the [c-ares resolver](https://github.com/envoyproxy/envoy/blob/db756afd204e186af609583badd3a4fb0a5a8178/source/common/network/dns_impl.cc#L161-L171) which catches other exception types and throws them as EnvoyExceptions (plus post the throw). But, yeah if we don't mind about the exception type I can just call the callback without the try.

Reply:
Given that we are calling code we own, I don't know why we would be catching other types and wrapping them, so I would just remove all of this.